libkbfs/dirty_bcache.go

```diff
@@ -20,14 +22,37 @@ type dirtyBlockID struct {
 type dirtyReq struct {
 	respChan chan<- struct{}
 	bytes    int64
+	start    time.Time
+	deadline time.Time
 }
 
 // DirtyBlockCacheStandard implements the DirtyBlockCache interface by
 // storing blocks in an in-memory cache.  Dirty blocks are identified
 // by their block ID, branch name, and reference nonce, since the same
 // block may be forked and modified on different branches and under
 // different references simultaneously.
+//
+// DirtyBlockCacheStandard controls how fast uses can write into KBFS,
+// and does so with a TCP-like slow-start algorithm that adjusts
+// itself according to how fast bytes are synced to the server.  It
+// keeps track of a "sync buffer" that holds all of the bytes that
+// have been written but not yet synced to the server.  When the
+// buffer is full, write requests get blocked by a fraction of the
+// write timeout, according to how close the user is to filling up
+// their "dirty buffer" -- i.e., the total number of bytes for which
+// an overall Sync operation has not yet completed (even if their
+// individual block Puts to the server have completed).
+//
+// The size of the sync buffer can vary between a minimum and maximum.
+// It starts out at the minimum, and grows exponentially as Puts
+// succeed.  However, if a write request is blocked for too long by
+// the backpressure and buffer-fullness checks, the sync buffer size
+// is cut in half.
```

Comment:
There is no decay currently.  I figure we can add one later if needed, but am open to doing it now if someone feels strongly that we should.